controllers/control_plane_init_locker.go

```diff
@@ -18,52 +18,56 @@ package controllers
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 
 	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
 	apicorev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	clusterv2 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// ControlPlaneInitLocker provides a locking mechanism for cluster initialization.
-type ControlPlaneInitLocker interface {
-	// Acquire returns true if it acquires the lock for the cluster.
-	Acquire(cluster *clusterv2.Cluster) bool
-}
+const lockInformationKey = "lock-information"
 
-// controlPlaneInitLocker uses a ConfigMap to synchronize cluster initialization.
-type controlPlaneInitLocker struct {
+// ControlPlaneInitLocker uses a ConfigMap to synchronize cluster initialization.
+type ControlPlaneInitLocker struct {
 	log    logr.Logger
 	client client.Client
-	ctx    context.Context
 }
 
-var _ ControlPlaneInitLocker = &controlPlaneInitLocker{}
-
-func newControlPlaneInitLocker(ctx context.Context, log logr.Logger, client client.Client) *controlPlaneInitLocker {
-	return &controlPlaneInitLocker{
-		ctx:    ctx,
+// NewControlPlaneInitLocker returns a lock that can be held by a control plane node before init.
+func NewControlPlaneInitLocker(log logr.Logger, client client.Client) *ControlPlaneInitLocker {
+	return &ControlPlaneInitLocker{
 		log:    log,
 		client: client,
 	}
 }
 
-func (l *controlPlaneInitLocker) Acquire(cluster *clusterv2.Cluster) bool {
-	configMapName := fmt.Sprintf("%s-controlplane", cluster.UID)
-	log := l.log.WithValues("namespace", cluster.Namespace, "cluster-name", cluster.Name, "configmap-name", configMapName)
+func lockName(clusterUID types.UID) string {
+	return fmt.Sprintf("%s-controlplane", clusterUID)
```

Comment:
When I was writing this for CAPA v1a1, I used the UID to avoid exceeding the max name length for metadata.name. But based on discussion at this week's meeting, we decided to use the cluster name as a prefix to related objects (e.g. secrets for kubeconfig, certificates, etc). And we'll eventually add a validating webhook that limits the max length of a cluster name so we can handle the various prefixes/suffixes (`controlplane`, `kubeconfig`, and so on). It's fine to leave this as a UID if you want, but it's also ok to switch it to the cluster name.

Reply:
If we switch to the clustername, we probably want to append `lock` to the name so that users don't mistaken believe it is something for them to use/edit.

Reply:
sounds like clustername-lock is the way of the future.